proposals/context.md

```diff
@@ -0,0 +1,265 @@
+
+```js
+this.dispatchEvent(
+  new RequestContextEvent(
+    "cool-thing-we-want-once",
+    (coolThing, dipose) => {
+      // if we were given a disposer, this provider is likely to send us updates
+      if (dispose) {
+        // so dispose immediately
+        dispose();
+      }
+      // guard against multiple assignment in case of bad actor providers
+      if (!this.myCoolThing) {
+        this.myCoolThing = coolThing; // do something with value
+      }
+    },
+    true // we only want the event once
+  )
+);
+```
+
+It is recommended that custom elements which participate in the Context API should fire their context-request events in their connectedCallback handler. Likewise in their disconnectedCallback they should invoke any dispose functions they have received.
+
+A more complete example is as follows:
+
+```js
+class SimpleElement extends HTMLElement {
+  connectedCallback() {
+    this.dispatchEvent(
+      new RequestContextEvent("logger", (value, dispose) => {
+        // protect against changing providers
+        if (dispose && dispose !== this.loggerDisposer) {
+          this.dispose();
+        }
+        this.logger = value;
+        this.loggerDisposer = dispose;
+      })
+    );
+  }
+  disconnectedCallback() {
+    if (this.loggerDisposer) {
+      this.loggerDisposer();
+    }
+    this.loggerDisposer = undefined;
+    this.logger = undefined;
+  }
+}
+```
+
+# Open Questions / Previous considerations
+
+## Could we use Promises instead of callbacks?
+
+Many have commented that Promises could be used instead of callback functions. The major drawback of promises is that they do not allow multiple-resolution. Therefore we would not be able to handle cases where a requested value changes over time. This is capability that we see in the React Context API and believe is valuable for a variety of use cases.
+
+While we could restrict this API to only support a single resolution of a requested value, and then use observable mechanisms on that value to achieve data update behaviors, it is believe this will complicate simple use-cases unnecessarily.
+
+## What about a `createContext` style API?
+
+In React Context API they provide a function `createContext` which creates a context object that uniquely identifies the context to consumers and providers. This has some advantages in that it means you have a unique object that identifies a context. However the downside is that you need to have an agreed upon module that exports this context value in order to share the context.
+
+For web components this creates a centralization issue. If all components want is a logger context, where would the logger context object be defined? Who owns that module? Will all consumers agree to depend upon it?
```

Comment:
If all requestors of a logger context are expecting a particular API, isn't it better if they indicate that by using a shared module? What happens if multiple consumers depend on incompatible logger interfaces keyed by the same context name?

Reply:
I agree its definitely more strict to work that way.

Can we hope for a consistent implementation though across the WC community for common contexts? I think if we could have the community-protocols repository produce some libraries of common contexts and their associated API shapes this might work well perhaps?

Reply:
After some offline discussion an outline of how this might look is as follows, new definitions

```typescript
export type Context<T> = {
    key: string | symbol;
    _type?: T; // flavored brand for the type system
};

export type AnyContext = Context<unknown>;

export function createContext<T>(key: string | symbol): Context<T> {
    return {
        key,
    };
}
```

These would then be used something like this to define a new context and export it:

```ts
// define an interface for the context value
export interface Logger {
    log: (msg: string) => void;
};
// create a context object
export const loggerContext = createContext<Logger>('logger');
```

Components, applications and libraries which produce and consume contexts should publish modules containing them, preferably small atomic modules with just interfaces and the context object. Perhaps common contexts could be agreed upon and interfaces published by groups like the community-protocols group.

The context could then be requested like so:

```ts
import { loggerContext, Logger } from '@community-protocols/contexts/logger'; // a common package

// later a component could get the context via event
this.dispatchEvent(new RequestContextEvent(loggerContext, (logger) => { this.logger = logger; }));
```

This has one major advantage over the string based keys in that it completely avoid namespace collisions. It also simplifies usage of the type system in Typescript avoiding the need for interface map merging.

I remain a little concerned that this does open the possibility for more and more functionality to be added to the context objects returned by `createContext` and that this might make compatibility harder over time as libraries adopt divergent implementations. But I do like the increased ease of type safety, and the name collision avoidance.

Reply:
Have updated the proposal to match the above in  66ec2ad